### DISCUSSION TITLE
style-elements-across-nested-layouts-with-tailwind

### DISCUSSION THREAD

**Discussion by timomeh** [timomeh/timomeh.de#174] [Posts]
~~~
cover_image: https://github.com/timomeh/timomeh.de/assets/4227520/5b9f737a-9dae-496b-abfe-19582adb8326
~~~

# Style elements across nested layouts with Tailwind

Tailwind has this really useful [`group`](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state) utility, that I recently discovered is _super useful_ in frameworks with nested layouts. 

You can use `group` to style any element based on **any other element** in the DOM. This may not seem like a big deal, but consider this example where we can hide the header if there's an element with the `hide-header-signal` class somewhere in the DOM.

```html
<body class="group/body">
  <header class="group-has-[.hide-header-signal]/body:hidden">...</header>
  <main>
    <h1>Hello world</h1>
    <div class="hide-header-signal"></div>
  </main>
</body>
```

This uses the [named group](https://tailwindcss.com/docs/hover-focus-and-other-states#differentiating-nested-groups) `group/body` to uniquely identify the group, in case we add more groups later.

You might think that you could just conditionally render the header to achieve the same thing, but that's not always easy to do, especially in frameworks with nested layouts and streaming responses, like Next.js.

Here's a real world example: This blog has a header on all pages, so the header can simply be in the root layout. But I also have two pages for visual regression testing of my [syntax highlighting](https://timomeh.de/vrt/shiki) and [markdown rendering](https://timomeh.de/vrt/markdown), and I don't want to show the header on those pages, because the tests shouldn't fail when I change the header.

- You don't have access to the current URL in the root layout, in order to conditionally hide the header based on the URL, unless you make the layout a client component – because of partial rendering and streaming.
- You could also move the header into every page instead of the layout, and not include it in the pages where you don't want to show the header, but then you don't take advantage of your layouts.
- You could create more nested folders with nested layouts, but you do it _just for those two pages_.

So although you could do a bunch of things to render the header dynamically for those pages, it has a comparatively high impact on our codebase. Sometimes this impact is justified and one of these approaches might be a better choice, but clearly not in this case where I just have some unlisted pages for testing purposes.

By using the pattern above, with our `group` utility friend, it's now super easy to hide the header on some pages:

```tsx
// layout.tsx
export default function RootLayout({ children }) {
  return (
    <html>
      <body className="group/body">
        <header className="group-has-[.hide-header-signal]/body:hidden">
          <Header />
        </header>
        {children}
      </body>
    </html>
  )
}

// e.g. regular blog post pages
// posts/[...slug]/page.tsx
export default function Page() {
  return (
    <main>
      <p>The header is visible on this page</p>
    </main>
  )
}

// e.g. pages for visual regression testing
// vrt/[...name]/page.tsx
export default function Page() {
  return (
    <main>
      <p>The header is hidden on this page</p>
      <div className="hide-header-signal" />
    </main>
  )
}
```

Done. A simple change which just added a few classes and affected 2 files, instead of changing a lot of code and affecting many files.

This example is only using the `has-[.hide-header-signal]` modifier, but it's not limited to only this modifier. You can do even more than to simply hide one element based on the existence of another element. For example, you could change the position of an element based on the data attribute of another element. Or change colors of the navigation – you know, sometimes websites are generally in light mode but then have some special pages where it's suddenly dark mode.

And the nice thing is: it's just CSS.

I guess these selectors are slow as hell, but get your shit together, browsers.

**Comment by H-Hp** [comment]
Hi
The `group` utility in Tailwind CSS is indeed a powerful feature that allows you to style elements based on the state or presence of other elements in the DOM. It's particularly useful in scenarios where you have nested layouts or complex component structures, and you need to apply styles across different levels of the hierarchy.

In your example, using the `group` utility to hide the header based on the presence of a `hide-header-signal` class is a clever and elegant solution. It avoids the need for conditional rendering or more complex layout structures, which can impact the codebase and make it harder to maintain.

You're right that using the `group` utility might come with a performance cost, as it relies on complex CSS selectors. However, modern browsers have become increasingly efficient at handling complex CSS selectors, and the performance impact might be negligible in most cases.

One possible alternative to the `group` utility, while still leveraging Tailwind CSS, is to use state management solutions like React Context or Jotai. You could create a global state that controls the visibility of the header, and pass that state down to the relevant components through context or props.

Here's an example of how you could achieve the same functionality using React Context:

```jsx
// HeaderContext.jsx
import React, { createContext, useState } from 'react';

export const HeaderContext = createContext(null);

export const HeaderProvider = ({ children }) => {
  const [hideHeader, setHideHeader] = useState(false);

  return (
    <HeaderContext.Provider value={{ hideHeader, setHideHeader }}>
      {children}
    </HeaderContext.Provider>
  );
};

// RootLayout.jsx
import { HeaderContext } from './HeaderContext';

export default function RootLayout({ children }) {
  return (
    <HeaderProvider>
      <html>
        <body>
          <header className={hideHeader ? 'hidden' : ''}>
            <Header />
          </header>
          {children}
        </body>
      </html>
    </HeaderProvider>
  );
}

// Page.jsx
import { useContext } from 'react';
import { HeaderContext } from './HeaderContext';

export default function Page() {
  const { setHideHeader } = useContext(HeaderContext);

  return (
    <main>
      <p>The header is hidden on this page</p>
      <button onClick={() => setHideHeader(true)}>Hide Header</button>
    </main>
  );
}
```

In this example, we create a `HeaderContext` that manages the state of the header visibility. The `RootLayout` component wraps the entire application with the `HeaderProvider`, which provides the `hideHeader` state and a function to update it (`setHideHeader`). In the `Page` component, we can import the `HeaderContext` and use the `setHideHeader` function to toggle the visibility of the header.

While this approach requires more code and complexity, it can potentially offer better performance, as it doesn't rely on complex CSS selectors. Additionally, it provides a centralized way to manage the visibility of the header, which can be beneficial as your application grows in complexity.

Ultimately, both approaches have their pros and cons, and the choice between them will depend on your specific requirements, performance considerations, and personal preferences.

**Reply by timomeh** [reply]
lol, thanks ChatGPT

using react context makes no sense here and hiding it on a button click makes even less sense, but good to know that an LLM is still dumb enough
